File: native/spatial_alc.c
/* Original spatial ALC: 8-frame stereo detector, 24-frame lookahead, Q31 log approximation. */
#include "spatial_alc.h"
#include <float.h>
#include <math.h>
#include <string.h>
static double native_log8(double value){
    static const int32_t table[12]={0,-186064426,-372130999,-558195425,-744261998,-930326424,-1116390849,-1302457422,-1488521848,-1674588421,-1860652847,-2046717273};
    int positive=0;
    while(value>=1.0){value*=.5;positive++;}
    int64_t wide=(int64_t)(value*2147483648.0);
    int32_t q=(int32_t)(wide>INT32_MAX?INT32_MAX:(wide<INT32_MIN?INT32_MIN:wide));
    int shift=0;
    while(q<0x40000000&&shift<11){q=(int32_t)((uint32_t)q*2u);shift++;}
    if(shift==11)q=INT32_MAX;
    int64_t x=(int32_t)((uint32_t)q+0x80000000u);
    int64_t square=(int32_t)((x*x)>>31);
    int64_t cube=(int32_t)((square*x)>>31);
    int64_t polynomial=((cube*0x2aaaaaaaLL)>>31)-(square>>1)+x;
    int32_t result=(int32_t)((uint32_t)(polynomial>>3)+(uint32_t)table[shift]);
    return (double)result*0x1p-31+(double)positive*0.08664339756999312;
}
void inzone_spatial_alc_init(InzoneSpatialAlc *s,int boost){memset(s,0,sizeof(*s));s->gain=pow(10.0,(double)boost/20.0);}
void inzone_spatial_alc_block(InzoneSpatialAlc *s,const float *input,float *output){
    double current[8][2],peak=0;
    for(int i=0;i<8;i++)for(int ch=0;ch<2;ch++){
        double x=(double)input[2*i+ch]*s->gain;
        if(fabs(x)<FLT_MIN)x=0;
        current[i][ch]=x;if(fabs(x)>peak)peak=fabs(x);
    }
    double change=peak-s->peak;
    if(change>0)s->peak+=change*((double)0x67d2ec9b*0x1p-31);
    else s->peak*=(double)0x7ac6b85a*0x1p-31;
    double logpeak=native_log8(s->peak);
    double gain=logpeak>0?exp(-logpeak*8):1.0;
    for(int i=0;i<8;i++)for(int ch=0;ch<2;ch++){
        double x=fmin(1.0,fmax(-1.0,s->history[i][ch]*gain));
        if(fabs(x)<FLT_MIN)x=0;
        output[2*i+ch]=(float)x;
    }
    memmove(s->history,s->history+8,16*sizeof(s->history[0]));
    memcpy(s->history+16,current,sizeof(current));
}

File: Sources/InzoneDSP/SpatialALC.swift
/// The detector preserves the recovered eight-frame cadence and 24-frame lookahead.
struct SpatialALCState {
    private var peak: Double = 0
    private var gain: Double = 1
    private var history = InlineArray<48, Double>(repeating: 0)

    init() {}

    @_noLocks
    mutating func reset(boost: Int) {
        peak = 0
        gain = dspPower(10.0, Double(boost) / 20.0)
        for index in 0..<48 {
            history[index] = 0
        }
    }

    @_noLocks
    mutating func process(input: UnsafePointer<Float>, output: UnsafeMutablePointer<Float>) {
        var current = InlineArray<16, Double>(repeating: 0)
        var currentPeak: Double = 0
        let minimumNormal = Double(Float.leastNormalMagnitude)
        for index in 0..<16 {
            var sample = Double(input[index]) * gain
            if Swift.abs(sample) < minimumNormal {
                sample = 0
            }
            current[index] = sample
            if Swift.abs(sample) > currentPeak {
                currentPeak = Swift.abs(sample)
            }
        }

        let change = currentPeak - peak
        if change > 0 {
            peak += change * (Double(0x67d2ec9b) * 0x1p-31)
        } else {
            peak *= Double(0x7ac6b85a) * 0x1p-31
        }
        let logarithmicPeak = Self.logarithmDividedByEight(peak)
        let currentGain = logarithmicPeak > 0 ? dspExponential(-logarithmicPeak * 8) : 1.0
        for index in 0..<16 {
            var sample = dspMinimum(1.0, dspMaximum(-1.0, history[index] * currentGain))
            if Swift.abs(sample) < minimumNormal {
                sample = 0
            }
            output[index] = Float(sample)
        }

        for index in 0..<32 {
            history[index] = history[index + 16]
        }
        for index in 0..<16 {
            history[index + 32] = current[index]
        }
    }

    @_noLocks
    private static func logarithmDividedByEight(_ input: Double) -> Double {
        var value = input
        var positiveShifts = 0
        while value >= 1.0 {
            value *= 0.5
            positiveShifts += 1
        }

        let wide = Int64(value * 2_147_483_648.0)
        var quantized = Int32(clamping: wide)
        var shift = 0
        while quantized < 0x40000000 && shift < 11 {
            quantized = Int32(bitPattern: UInt32(bitPattern: quantized) &* 2)
            shift += 1
        }
        if shift == 11 {
            quantized = Int32.max
        }

        // Q31 intermediates retain the original signed truncation and unsigned wraparound.
        let shifted = Int64(Int32(bitPattern: UInt32(bitPattern: quantized) &+ 0x80000000))
        let square = Int64(Int32(truncatingIfNeeded: (shifted * shifted) >> 31))
        let cube = Int64(Int32(truncatingIfNeeded: (square * shifted) >> 31))
        let polynomial = ((cube * 0x2aaaaaaa) >> 31) - (square >> 1) + shifted
        let result = Int32(bitPattern:
            UInt32(truncatingIfNeeded: polynomial >> 3)
                &+ UInt32(bitPattern: logarithmOffset(shift)))
        return Double(result) * 0x1p-31 + Double(positiveShifts) * 0.08664339756999312
    }

    @_noLocks
    private static func logarithmOffset(_ shift: Int) -> Int32 {
        // A switch keeps the fixed table free of lazy initialization in the audio callback.
        switch shift {
        case 0: return 0
        case 1: return -186_064_426
        case 2: return -372_130_999
        case 3: return -558_195_425
        case 4: return -744_261_998
        case 5: return -930_326_424
        case 6: return -1_116_390_849
        case 7: return -1_302_457_422
        case 8: return -1_488_521_848
        case 9: return -1_674_588_421
        case 10: return -1_860_652_847
        default: return -2_046_717_273
        }
    }
}
